examples/graph/sdpa_bottom_right_causal_mask.cpp

```diff
@@ -0,0 +1,360 @@
+    const dims k_sz = {p.mb, p.head_num, p.seq_len, p.head_size};
+    const dims score_sz = {p.mb, p.head_num, p.query_num, p.seq_len};
+    const dims scale_sz = {1};
+
+    // Incremental IDs used to create logical tensors and operations.
+    size_t id = 0;
+
+    // Intermediate data type
+    const logical_tensor::data_type dt_inter = logical_tensor::data_type::f32;
+
+    // score = query x key.T
+    auto query = logical_tensor(id++, dt, qv_sz, layout_type::strided);
+    auto key = logical_tensor(id++, dt, k_sz, layout_type::strided);
+    auto score = logical_tensor(id++, dt_inter, score_sz, layout_type::strided);
+    auto bmm1 = op(id++, op::kind::MatMul, "bmm1");
+    bmm1.set_attr<bool>(op::attr::transpose_b, true);
+    bmm1.add_inputs({query, key});
+    bmm1.add_outputs({score});
+
+    // scale_mul_out_lt = score * scale
+    auto scale = logical_tensor(id++, dt, scale_sz, layout_type::strided);
+    auto scaled_score
+            = logical_tensor(id++, dt_inter, score_sz, layout_type::strided);
+    auto scale_mul = op(id++, op::kind::Multiply, "scale_mul");
+    scale_mul.add_inputs({score, scale});
+    scale_mul.add_outputs({scaled_score});
+
+    auto index_row = logical_tensor(
+            id++, data_type::s32, score_sz, layout_type::strided);
+    auto gen_index_row = op(id++, op::kind::GenIndex, "gen_index_row");
+    gen_index_row.set_attr<int64_t>(op::attr::axis, -2);
+    gen_index_row.add_inputs({scaled_score});
+    gen_index_row.add_outputs({index_row});
+
+    auto seq_len_kv = logical_tensor(id++, data_type::s32, 0,
+            layout_type::strided, property_type::host_scalar);
+    auto mask_add_out = logical_tensor(
+            id++, data_type::s32, score_sz, layout_type::strided);
+    auto mask_add = op(id++, op::kind::Add, "mask_add");
+    mask_add.add_inputs({index_row, seq_len_kv});
+    mask_add.add_outputs({mask_add_out});
+
+    auto seq_len_q = logical_tensor(id++, data_type::s32, 0,
+            layout_type::strided, property_type::host_scalar);
+    auto mask_sub_out = logical_tensor(
+            id++, data_type::s32, score_sz, layout_type::strided);
+    auto mask_sub = op(id++, op::kind::Subtract, "mask_sub");
+    mask_sub.add_inputs({mask_add_out, seq_len_q});
+    mask_sub.add_outputs({mask_sub_out});
+
+    auto index_col = logical_tensor(
+            id++, data_type::s32, score_sz, layout_type::strided);
+    auto gen_index_col = op(id++, op::kind::GenIndex, "gen_index_col");
+    gen_index_col.set_attr<int64_t>(op::attr::axis, -1);
+    gen_index_col.add_inputs({scaled_score});
+    gen_index_col.add_outputs({index_col});
+
+    auto mask_ge_out = logical_tensor(
+            id++, data_type::boolean, score_sz, layout_type::strided);
+    auto mask_ge = op(id++, op::kind::GreaterEqual, "mask_ge");
+    mask_ge.add_inputs({mask_sub_out, index_col});
+    mask_ge.add_outputs({mask_ge_out});
+
+    auto neg_inf
+            = logical_tensor(id++, dt_inter, scale_sz, layout_type::strided);
```

Comment:
Also use host_scalar for negative infinity?

Reply:
No. Let's add negative infinity support along with the scalar attention.